src/Common/tests/Tests/System/IO/PathInternal.Tests.cs

```diff
@@ -64,5 +64,220 @@ public void GetCommonPathLength(string first, string second, bool ignoreCase, in
         {
             Assert.Equal(expected, PathInternal.GetCommonPathLength(first, second, ignoreCase));
         }
+
+        public static TheoryData<string, int, string> RemoveRelativeSegmentsData => new TheoryData<string, int, string>
+        {
+            { @"C:\git\corefx", 2, @"C:\git\corefx"},
+            { @"C:\\git\corefx", 2, @"C:\git\corefx"},
+            { @"C:\git\\corefx", 2, @"C:\git\corefx"},
+            { @"C:\git\.\corefx\.\\", 2, @"C:\git\corefx\"},
+            { @"C:\git\corefx", 2, @"C:\git\corefx"},
+            { @"C:\git\..\corefx", 2, @"C:\corefx"},
+            { @"C:\git\corefx\..\", 2, @"C:\git\"},
+            { @"C:\git\corefx\..\..\..\", 2, @"C:\"},
+            { @"C:\git\corefx\..\..\.\", 2, @"C:\"},
+            { @"C:\git\..\.\corefx\temp\..", 2, @"C:\corefx"},
+            { @"C:\git\..\\\.\..\corefx", 2, @"C:\corefx"},
+            { @"C:\git\corefx\", 2, @"C:\git\corefx\"},
+            { @"C:\git\temp\..\corefx\", 2, @"C:\git\corefx\"},
+        };
+
+        public static TheoryData<string, int, string> RemoveRelativeSegmentsFirstRelativeSegment => new TheoryData<string, int, string>
+        {
+            { @"C:\\git\corefx", 2, @"C:\git\corefx"},
+            { @"C:\.\git\corefx", 2, @"C:\git\corefx"},
+            { @"C:\\.\git\.\corefx", 2, @"C:\git\corefx"},
+            { @"C:\..\git\corefx", 2, @"C:\git\corefx"},
+            { @"C:\.\git\..\corefx", 2, @"C:\corefx"},
+            { @"C:\.\git\corefx\..\", 2, @"C:\git\"},
+            { @"C:\.\git\corefx\..\..\..\", 2, @"C:\"},
+            { @"C:\.\git\corefx\..\..\.\", 2, @"C:\"},
+            { @"C:\.\git\..\.\corefx\temp\..", 2, @"C:\corefx"},
+            { @"C:\.\git\..\\\.\..\corefx", 2, @"C:\corefx"},
+            { @"C:\.\git\corefx\", 2, @"C:\git\corefx\"},
+            { @"C:\.\git\temp\..\corefx\", 2, @"C:\git\corefx\"},
+        };
+
+        public static TheoryData<string, int, string> RemoveRelativeSegmentsFirstRelativeSegmentRoot => new TheoryData<string, int, string>
+        {
+            { @"C:\\git\corefx", 3, @"C:\\git\corefx"},
+            { @"C:\.\git\corefx", 3, @"C:\.\git\corefx"},
+            { @"C:\\.\git\.\corefx", 3, @"C:\\git\corefx"},
+            { @"C:\..\git\corefx", 3, @"C:\..\git\corefx"},
```

Comment:
@JeremyKuhne what should we do in cases like  when the skip is between the directory name
```
{ @"C:\tmp\..\..\..\", 5, @"C:\tmp\" },
{ @"C:\tmp\..", 5, @"C:\tmp" }
```

Reply:
should we keep the segment or remove it ?

Reply:
I don't really care about that one. In order for `C:` to work correctly we can't really determine the segment super easy. I wouldn't introduce complexity for it as our root logic would be wrong.

Reply:
We require this for a particular case in out GetFullPath case https://github.com/dotnet/corefx/pull/27085/files#diff-c0b5df44050f78f48b85de7f90541bfaR258
As we are sending skip as 
```
            // Device paths are normalized by definition, so passing something of this format
            // to GetFullPath() won't do anything by design. Additionally, GetFullPathName() in
            // Windows doesn't root them properly. As such we need to manually remove segments.
            return PathInternal.IsDevice(combinedPath)
                // Paths at this point are in the form of \\?\C:\.\tmp we skip to the last character of the root when calling RemoveRelativeSegments to remove relative paths in such cases.
                ? PathInternal.RemoveRelativeSegments(combinedPath, PathInternal.GetRootLength(combinedPath) - 1)
                : GetFullPath(combinedPath);
```
I suggest the changes as https://github.com/dotnet/coreclr/pull/16869/commits/bef4597783bce9c2c967695fec12c172d28e6e4b
It passes all the current tests. Otherwise we will have to change GetFullPath(string, string) to correct this option

Reply:
That change you link to desperately needs a comment. Please explain what you are doing and why.

At first glance it seems fundamentally wrong to have to tweak what GetRootLength returns before you pass it to remove relative segments. This API should just work with the results that you get out of GetRootLength.

This seems disconnected and redundant.

``` C#
// Paths at this point are in the form of \\?\C:\.\tmp we skip to the last character of the root when calling RemoveRelativeSegments to remove relative paths in such cases.
```

It is much easier to read inline with the existing comment. I've tweaked it below:

``` C#
// Device paths are normalized by definition, so passing something of this format (i.e. \\?\C:\.\tmp, \\.\C:\foo)
// to Windows APIs won't do anything by design. Additionally, GetFullPathName() in Windows doesn't root
// them properly. As such we need to manually remove segments and not use GetFullPath().
return PathInternal.IsDevice(combinedPath)
    ? PathInternal.RemoveRelativeSegments(combinedPath, PathInternal.GetRootLength(combinedPath))
    : GetFullPath(combinedPath);
```

If you need to tweak the length I would call it out explicitly in the comments:

``` C#
// Device paths are normalized by definition, so passing something of this format (i.e. \\?\C:\.\tmp, \\.\C:\foo)
// to Windows APIs won't do anything by design. Additionally, GetFullPathName() in Windows doesn't root
// them properly. As such we need to manually remove segments and not use GetFullPath(). Note that we need
// to subtract one from GetRootLength because...
```